core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.querystats;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.stats.TimeDistribution;
+import io.airlift.stats.TimeDistribution.TimeDistributionSnapshot;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRuleStats
+{
+    private final String rule;
+    private long invocations;
+    private long applications;
+    private final TimeDistribution time = new TimeDistribution(TimeUnit.MICROSECONDS);
+    private long totalTime;
+    private long failures;
+
+    public QueryRuleStats(String rule)
+    {
+        this.rule = rule;
+    }
+
+    public void record(long nanos, boolean match)
+    {
+        if (match) {
+            applications += 1;
+        }
+
+        invocations += 1;
+        totalTime += nanos;
+        time.add(nanos);
+    }
+
+    public void recordFailure()
+    {
+        failures += 1;
+    }
+
+    @JsonProperty
+    public String getRule()
+    {
+        return rule;
+    }
+
+    @JsonProperty
+    public long getInvocations()
+    {
+        return invocations;
+    }
+
+    public TimeDistributionSnapshot getTime()
+    {
+        return time.snapshot();
+    }
+
+    public long getApplications()
+    {
+        return applications;
+    }
+
+    public long getFailures()
+    {
+        return failures;
+    }
+
+    public long getTotalTime()
+    {
+        return totalTime;
+    }
+
+    public QueryRuleStatsSnapshot snapshot()
+    {
+        return new QueryRuleStatsSnapshot(rule, invocations, applications, time.snapshot(), totalTime, failures);
+    }
+
+    public static class QueryRuleStatsSnapshot
```

Comment:
We can move this class to `io.trino.spi.eventlistener` and expose it to the event listener by adding the top-n list of these to the QueryStatistics class.

Reply:
SPI does not have `TimeDistributionSnapshot` class in dependencies.

Reply:
We can add a `TimeDistribution` to `io.trino.spi.eventlistener`, it's more convenient for the event listener to have visibility to the class used to serialize the json rather than a raw json string.
Feel free to do all the event listener changes as a follow up.